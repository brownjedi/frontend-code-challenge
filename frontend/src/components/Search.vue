<template>
  <div :class="styles.container">
    <img src="/img/search.svg" :class="styles['search-icon']" />
    <input
      type="text"
      ref="search"
      data-cy="pokemon-search"
      v-model="text"
      :class="styles.search"
      placeholder="Search PokÃ©mons"
    />
    <img
      src="/img/close.svg"
      tabindex="0"
      :class="styles['close-icon']"
      @click="onClose"
      @keyup.enter="onClose"
      @keydown.space.prevent="onClose"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, useCssModule, watch } from 'vue'
import { useDebounce } from '@vueuse/core'

export default defineComponent({
  props: {
    initialValue: {
      type: String,
    },
  },
  emits: ['onSearch'],
  setup(props, { emit }) {
    const search = ref(null)
    const text = ref(props.initialValue)
    const debounced = useDebounce(text, 600)
    const styles = useCssModule()

    function onClose() {
      text.value = ''
      const el = search.value as unknown as HTMLInputElement
      el.focus()
    }

    watch(debounced, value => emit('onSearch', value))

    return { styles, search, text, onClose }
  },
})
</script>

<style lang="scss" module>
.container {
  position: relative;
  display: flex;
  min-width: 18rem;
  height: 2rem;
  align-items: center;
  max-width: 60rem;
  margin: 0 auto;
}

.search {
  width: 100%;
  height: 100%;
  border-radius: 0.25rem;
  border: 2px solid var(--number-title-color);
  padding-left: 2rem;
  outline-color: var(--main-text-color);
}

.search-icon {
  width: 1.25rem;
  height: 1.25rem;
  display: block;
  position: absolute;
  left: 0.625rem;
}

.close-icon {
  width: 1.25rem;
  height: 1.25rem;
  display: block;
  position: absolute;
  right: 0.625rem;
  cursor: pointer;
  visibility: hidden;
}

.search:not(:placeholder-shown) + .close-icon {
  visibility: visible;
}
</style>
