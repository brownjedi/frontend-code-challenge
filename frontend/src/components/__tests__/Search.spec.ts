import { shallowMount } from '@vue/test-utils'
import Search from '@/components/Search.vue'

describe('components/Search.vue', () => {
  it('has the expected html structure', () => {
    const wrapper = shallowMount(Search)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('the input element works correctly', async () => {
    const sampleText = 'pikachu'
    const wrapper = shallowMount(Search)
    const input = wrapper.find('input')
    await input.setValue(sampleText)

    expect(wrapper.vm.text).toBe(sampleText)
    expect(input.element.value).toBe(sampleText)
  })

  it('the initial value is set correctly', async () => {
    const sampleText = 'pikachu'
    const wrapper = shallowMount(Search, {
      props: {
        initialValue: sampleText,
      },
    })
    const input = wrapper.find('input')
    expect(input.element.value).toBe(sampleText)
  })

  it('the close element works correctly', async () => {
    const sampleText = 'pikachu'
    const wrapper = shallowMount(Search, {
      props: {
        initialValue: sampleText,
      },
    })

    expect(wrapper.find('input').element.value).toBe(sampleText)

    await wrapper.get('.close-icon').trigger('click')
    expect(wrapper.find('input').element.value).toBe('')
  })
})
