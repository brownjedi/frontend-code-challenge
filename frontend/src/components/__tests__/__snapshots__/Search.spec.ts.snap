// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Search.vue has the expected html structure 1`] = `
<div
  class="container"
>
  <img
    class="search-icon"
    src="/img/search.svg"
  />
  <input
    class="search"
    data-cy="pokemon-search"
    placeholder="Search PokÃ©mons"
    type="text"
  />
  <img
    class="close-icon"
    src="/img/close.svg"
    tabindex="0"
  />
</div>
`;
